realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -146,6 +151,23 @@ public boolean handleMessage(Message message) {
         return true;
     }
 
+    /**
+     * Properly handles when an async transaction completes. This will be treated as a REALM_CHANGED event when
+     * determining which queries to re-run and when to notify listeners.
+     * <p>
+     * NOTE: This is needed as it is not possible to combine a `Message.what` value and a callback runnable. So instead
+     * of posting two messages, we post a runnable that runs this method. This means it is possible to interpret
+     * `REALM_CHANGED + Runnable` as one atomic message.
+     *
+     * @param onSuccess onSuccess callback to run for the async transaction that completed.
+     */
+    public void handleAsyncTransactionCompleted(Runnable onSuccess) {
+        if (onSuccess != null) {
+            pendingOnSuccessAsyncTransactionCallbacks.add(onSuccess);
+        }
+        realmChanged(false);
```

Comment:
Good point, and that is actually an argument against trying to add callbacks from outside the message queue. Trying to reason about where to lock will increase the complexity even further. Not to mention that the current HandlerController currently separate change notifications from advance_read. That would also have be even more tightly coupled. 

I am starting to think that the implementation in this PR is actually better as it separates the concerns in a nice way. Also, the example you gave is not entirely accurate. It will put a lot of tasks in the RealmThreadExecutor for sure, but events will still only trickle through as each async job is completed, so I don't think we need to worry about pushing to many events to the message queue.

What do you think @beeender ?